### PR TITLE
Enrich link formatting

### PR DESCRIPTION
**Pull request by LoLei**
* Link normal URLs
* Link (/)r/\<subreddits>

### CODE REVIEW

File: frontend/components/Message.tsx [hunk 1]
```diff
@@ -1,9 +1,57 @@
...
+      const hrefWithoutEndChar = splitToTextWithoutTrailingAndTrailing(href)[0];
+      return formatLinkWithEndChar(hrefWithoutEndChar, clickableWithoutEndChar, endChar);
+    }
+    return formatLinkNormally(href, clickable);
+  }
+
+  /** 
+   * If a message contains URLs or (/)r/<subreddit> those parts will be linked.
+   */
+  function linkifyMessageContent(message: string): JSX.Element[] {
```
Comment (maxaudron) [inline frontend/components/Message.tsx:9]: Is the function run on every line? Could it be better if it's only run on lines that actually contain links?

I'm not sure about the performance impact but I smell more potential for funky shit to happen if it just splits all the messages.

File: frontend/components/Message.tsx [hunk 2]
```diff
@@ -1,9 +1,57 @@
...
+    return message.split(' ')
+      .map(messagePart => {
+        if (messagePart.startsWith('r/')) {
+          return formatLink('https://reddit.com/' + messagePart, messagePart);
+        } else if (messagePart.startsWith('/r/')) {
+          return formatLink('https://reddit.com' + messagePart, messagePart);
+        } else if (messagePart.startsWith('https://')) {
```
Comment (maxaudron) [inline frontend/components/Message.tsx]: you should match on the general scheme `://` to link everything and not just https. Have a look at the URI rfc https://tools.ietf.org/html/rfc3986#page-7

File: frontend/util/stringUtils.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+export function hasTrailingChar(text: string): boolean {
+  return (text.endsWith(',') || text.endsWith('.'));
```
Comment (maxaudron) [inline frontend/util/stringUtils.ts]: Should probably also match to other invalid chars like !, more dificult would be the ? as that is also valid in urls to start a query argument

File: frontend/components/Message.tsx (hunk 2 above)
Comment (LoLei) [reply frontend/components/Message.tsx]: I deliberately only chose to use `https://` so we don't link to insecure `http` URLs, also didn't want to bother with prepending the protocol to `www` or whatever. I also deemed linking email addresses and phone numbers etc. out of scope and questionably useful at best.

File: frontend/components/Message.tsx (hunk 1 above)
Comment (LoLei) [reply frontend/components/Message.tsx:9]: This method is now only run on messages that contain parts that are to be linked.

File: frontend/util/stringUtils.ts (hunk 3 above)
Comment (LoLei) [reply frontend/util/stringUtils.ts]: Added `!` and `?`. Who *ends* a URL with `?` anyway? .shrug

File: frontend/components/Message.tsx (hunk 2 above)
Comment (maxaudron) [reply frontend/components/Message.tsx]: leaving out the mailto: etc urls is fine to me, they're not so clear to parse anyways, but I'd at least support the `<schema>:// ` ones in a general manner, caring about the protocol they are is not our job imo

File: frontend/util/stringUtils.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+export function hasTrailingChar(text: string): boolean {
+  return [',', '.', '!', '?'].includes(text.slice(text.length - 1, text.length))
```
Comment (LoLei) [inline frontend/util/stringUtils.ts]: TODO: Remove that unnecessary slice

File: frontend/components/Message.tsx [hunk 5]
```diff
@@ -1,9 +1,64 @@
...
+          return formatLink(messagePart, messagePart);
+        } else {
+          return <span> {messagePart}</span>
+        }
+      });
+  }
+
+  function linkifyMessageContentIfMessageContainsUrlOrSubreddit(message: string): string | JSX.Element[] {
+    if (/^.*((https:\/\/)|(\/?r)).*$/.test(message)) {
```
Comment (LoLei) [inline frontend/components/Message.tsx]: TODO: Fix regex (Apparently matches anything with `r`)

**Review by maxaudron — APPROVED** [review]

**Pull request merged**